zerver/views/streams.py

```diff
@@ -580,6 +582,137 @@ def you_were_just_subscribed_message(
 RETENTION_DEFAULT: str | int = "realm_default"
 
 
+@transaction.atomic(savepoint=False)
+@require_non_guest_user
+@typed_endpoint
+def add_channel(
+    request: HttpRequest,
+    user_profile: UserProfile,
+    *,
+    channel: Annotated[Json[AddSubscriptionData], ApiParamConfig("channel")],
+    invite_only: Json[bool] = False,
+    is_web_public: Json[bool] = False,
+    is_default_stream: Json[bool] = False,
+    history_public_to_subscribers: Json[bool] | None = None,
+    message_retention_days: Json[str] | Json[int] = RETENTION_DEFAULT,
+    can_add_subscribers_group: Json[int | AnonymousSettingGroupDict] | None = None,
+    can_administer_channel_group: Json[int | AnonymousSettingGroupDict] | None = None,
+    can_send_message_group: Json[int | AnonymousSettingGroupDict] | None = None,
+    can_remove_subscribers_group: Json[int | AnonymousSettingGroupDict] | None = None,
+    subscribers: Json[list[str] | list[int]],
+    announce: Json[bool] = False,
+) -> HttpResponse:
+    realm = user_profile.realm
+    color_map = {}
+    setting_groups_dict = {}
+    group_settings_map = {}
+    request_settings_dict = locals()
+    # We don't want to calculate this value if no default values are
+    # needed.
+    system_groups_name_dict = None
+    for setting_name, permission_configuration in Stream.stream_permission_group_settings.items():
+        assert setting_name in request_settings_dict
+        if request_settings_dict[setting_name] is not None:
+            setting_request_value = request_settings_dict[setting_name]
+            setting_value = parse_group_setting_value(setting_request_value)
+            group_settings_map[setting_name] = access_user_group_for_setting(
+                setting_value,
+                user_profile,
+                setting_name=setting_name,
+                permission_configuration=permission_configuration,
+            )
+            setting_groups_dict[group_settings_map[setting_name].id] = setting_value
+        else:
+            if system_groups_name_dict is None:
+                system_groups_name_dict = get_role_based_system_groups_dict(realm)
+            group_settings_map[setting_name] = get_stream_permission_default_group(
+                setting_name, system_groups_name_dict, creator=user_profile
+            )
+            if permission_configuration.default_group_name == "stream_creator_or_nobody":
+                # Default for some settings like "can_administer_channel_group"
+                # is anonymous group with stream creator.
+                setting_groups_dict[group_settings_map[setting_name].id] = (
+                    AnonymousSettingGroupDict(direct_subgroups=[], direct_members=[user_profile.id])
+                )
+            else:
+                setting_groups_dict[group_settings_map[setting_name].id] = group_settings_map[
+                    setting_name
+                ].id
+
+    ensure_stream_does_not_exist_already(channel.name, user_profile.realm)
```

Comment:
This check should be super inexpensive, so I'm not sure why you don't lift it toward the top of the function.